File: features/list/src/test/java/com/vp/list/viewmodel/ListViewModelTest.java
package com.vp.list.viewmodel;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.Observer;
import com.vp.list.model.SearchResponse;
import com.vp.list.service.SearchService;
import java.io.IOException;
import java.util.List;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import retrofit2.mock.Calls;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class ListViewModelTest {
    @Rule
    public InstantTaskExecutorRule instantTaskRule = new InstantTaskExecutorRule();

    @Test
    public void shouldReturnErrorState() {
        //given
        SearchService searchService = mock(SearchService.class);
        when(searchService.search(anyString(), anyInt())).thenReturn(Calls.failure(new IOException()));
        ListViewModel listViewModel = new ListViewModel(searchService);

        //when
        listViewModel.searchMoviesByTitle("title", 1);

        //then
        assertThat(listViewModel.observeMovies().getValue().getListState()).isEqualTo(ListState.ERROR);
    }

    @Test
    public void shouldReturnInProgressState() {
        //given
        SearchService searchService = mock(SearchService.class);
        when(searchService.search(anyString(), anyInt())).thenReturn(Calls.response(mock(SearchResponse.class)));
        ListViewModel listViewModel = new ListViewModel(searchService);
        Observer<SearchResult> mockObserver = (Observer<SearchResult>) mock(Observer.class);
        listViewModel.observeMovies().observeForever(mockObserver);

        //when
        listViewModel.searchMoviesByTitle("title", 1);

        //then
        verify(mockObserver).onChanged(SearchResult.inProgress());
    }

    @Test
    public void shouldReturnOnSuccessState() {
        //given
        SearchService searchService = mock(SearchService.class);
        when(searchService.search(anyString(), anyInt())).thenReturn(Calls.response(mock(SearchResponse.class)));
        ListViewModel listViewModel = new ListViewModel(searchService);
        Observer<SearchResult> mockObserver = (Observer<SearchResult>) mock(Observer.class);
        listViewModel.observeMovies().observeForever(mockObserver);

        //when
        listViewModel.searchMoviesByTitle("title", 1);

        //then
        ArgumentCaptor<SearchResult> captor = ArgumentCaptor.forClass(SearchResult.class);
        verify(mockObserver, times(2)).onChanged(captor.capture());
        List<SearchResult> captorValues = captor.getAllValues();
        assertThat(captorValues.get(0).equals(SearchResult.inProgress()));
        assertThat(captorValues.get(1).equals(SearchResult.success(captorValues.get(1).getItems(), captorValues.get(1).getItems().size())));
    }

}

File: features/list/src/test/java/com/vp/list/viewmodel/ListViewModelTest.kt
package com.vp.list.viewmodel

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.Observer
import com.vp.list.model.SearchResponse
import com.vp.list.service.SearchService
import com.vp.list.viewmodel.ListState.ERROR
import com.vp.list.viewmodel.SearchResult.Companion.inProgress
import com.vp.list.viewmodel.SearchResult.Companion.success
import org.assertj.core.api.Assertions
import org.junit.Rule
import org.junit.Test
import org.mockito.Mockito
import org.mockito.Mockito.anyInt
import org.mockito.Mockito.anyString
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import retrofit2.mock.Calls
import java.io.IOException

fun <T> whenever(methodCall: T) = Mockito.`when`(methodCall)

class ListViewModelTest {
    @get:Rule
    var instantTaskRule = InstantTaskExecutorRule()

    @Test
    fun shouldReturnErrorState() {
        //given
        val searchService = mock(SearchService::class.java)
        whenever(searchService.search(anyString(), anyInt()))
            .thenReturn(Calls.failure(IOException()))
        val listViewModel = ListViewModel(searchService)

        //when
        listViewModel.searchMoviesByTitle("title", 1)

        //then
        Assertions.assertThat(listViewModel.movies.value!!.listState).isEqualTo(ERROR)
    }

    @Test fun shouldReturnInProgressState() {
        //given
        val searchService = mock(SearchService::class.java)
        whenever(searchService.search(anyString(), anyInt()))
            .thenReturn(Calls.response(mock(SearchResponse::class.java)))
        val listViewModel = ListViewModel(searchService)
        val mockObserver = mock(
            Observer::class.java
        ) as Observer<SearchResult>
        listViewModel.movies.observeForever(mockObserver)

        //when
        listViewModel.searchMoviesByTitle("title", 1)

        //then
        verify(mockObserver).onChanged(inProgress())
    }

    @Test fun shouldReturnInSuccessState() {
        //given
        val searchService = mock(SearchService::class.java)
        val searchResponse = mock(SearchResponse::class.java)
        whenever(searchService.search(anyString(), anyInt()))
            .thenReturn(Calls.response(searchResponse))
        val listViewModel = ListViewModel(searchService)
        val mockObserver = mock(
            Observer::class.java
        ) as Observer<SearchResult>
        listViewModel.movies.observeForever(mockObserver)

        //when
        listViewModel.searchMoviesByTitle("title", 1)

        //then
        verify(mockObserver).onChanged(inProgress())
        verify(mockObserver)
            .onChanged(success(searchResponse.search, searchResponse.totalResults))
    }
}
